docs(backups): encryption algorithms

thomas-dkmt opened the pull request:
Explain how backups are encrypted by Xen Orchestra, what each encryption algorithm (AES-256-GCM and ChaCha20-Poly1305) does, and how it works with Amazon S3.

thomas-dkmt committed:
backups.md: Explain backup encryption

Explain how backups are encrypted by Xen Orchestra,
what each encryption algorithm does,
and how it works with Amazon S3

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

julien-f committed:
format with Prettier

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
 
+## Backup Encryption
+
+Xen Orchestra ensures robust data security for backups stored in Amazon S3 by leveraging advanced encryption algorithms. Here’s a closer look at how encryption works and the technology behind it:
```
Encryption is compatible will all remotes, not only Amazon S3.

Encryption is opt-in and requires configuring an encryption key on the remote (page _Settings/Remotes_), I think we should show that.

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/6587b3a0f9e384f5078433d77ac0d8e00724fbd4

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+Currently, backups use the `AES-256-GCM` encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn’t an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform.
+
+In addition, this algorithm is fully compliant with [ANSSI guidelines](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```
```suggestion
In addition, this algorithm is fully compliant with [ANSSI guidelines 🇫🇷](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/commit/e4be7238a4ab0da40c52f8db74497f96788a9535.

Replaced the flag with a 'French' label. Flags aren't good practice, as per https://www.flagsarenotlanguages.com/

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+### Current Encryption Algorithm: `AES-256-GCM`
+
+Currently, backups use the `AES-256-GCM` encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn’t an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform.
```
```suggestion
Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn’t an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform.
```

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/259124a697cbda4162b4ad9daf33216e46186a35

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+To improve flexibility and performance, Xen Orchestra will transition to the `ChaCha20-Poly1305` encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
+
+Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.
```
```suggestion
Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.

This algorithm is also fully compliant with [ANSSI guidelines 🇫🇷](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/e4be7238a4ab0da40c52f8db74497f96788a9535.

Replaced the flag with a 'French' label. Flags aren't good practice, as per https://www.flagsarenotlanguages.com

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+### Upcoming Change: `ChaCha20-Poly1305`
+
+To improve flexibility and performance, Xen Orchestra will transition to the `ChaCha20-Poly1305` encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
```
```suggestion
To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
```

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/259124a697cbda4162b4ad9daf33216e46186a35

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,43 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+To encrypt the message, the cipher combines each bit of the message with the corresponding bit of the keystream using a simple XOR operation (a basic "bit-flipping" process).
+
+The result is the encrypted message, or ciphertext. Modern stream ciphers are designed to ensure the keystream appears completely random, even if someone tries to manipulate the initialization vector.
```
While it is quite interesting, I don't think it makes sense to put in our documentation. :thinking:

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/afcdbd80c20608e4585b4fdf9df041c2c268423b

thomas-dkmt committed:
backups.md: Remove unnecessary explanation of encryption algo

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
backups.md: Explain how to set up backup encryption

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
Merge branch 'doc-encryption-details' of https://github.com/vatesfr/xen-orchestra into doc-encryption-details

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
backups.md: Mention that the ANSSI link is in French

Note: using flags to identify languages is bad practice
Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
backups.md: Add links to explain the encryption algorithms

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

fbeauchamp commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,30 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+2. Go to the section called **New file system remote**, or edit an existing remote.
+3. In the subsection called **Encrypt all new data sent to this remote** you will find a text area. Enter your encryption key there.
+4. Click the **Save configuration** Button to finish the encryption setup.
+
```
encryption is only compatible with block based remote. (Full backup don't care, but incremental backup will fail if not )

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

fbeauchamp commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,30 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+Encryption is opt-in and requires configuring an encryption key on the remote.
+
+1. Go to the Settings → Remote menu.
+2. Go to the section called **New file system remote**, or edit an existing remote.
```
encryption can't be changed if a remote contains any backup (either enabling/disabling of changing the key )

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

fbeauchamp commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,30 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+### Current Encryption Algorithm: `AES-256-GCM`
+Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform. 
+
+In addition, this algorithm is fully compliant with [ANSSI guidelines (in French)](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
```
this algorithm is authenticated, that means an additional metadata is put at the end of any saved file. This is used on restore to ensure that the data restored are exactly the data that were encrypted, thus detecting bit rot or a malicious attacker without the key. Note this is not a recoverable error, the file will be unusable
https://en.wikipedia.org/wiki/Authenticated_encryption

thomas-dkmt replied:
Fixed by https://github.com/vatesfr/xen-orchestra/pull/8266/commits/4d784b0af9d1d060604fd4cde3ed8f00286342ee

thomas-dkmt replied:
~~TODO : it's the same for the other algorithm~~

fbeauchamp commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,30 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
+
+Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.
```
as aes-256-gcm , chacha20-poly1305 is authenticated

fbeauchamp commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,30 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+4. Click the **Save configuration** Button to finish the encryption setup.
+
+### Current Encryption Algorithm: `AES-256-GCM`
+Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with Amazon S3, as the data is split into smaller blocks, making it fully compatible with the platform. 
```
```suggestion
Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with incremental backups, as the data is split into smaller blocks, making it fully compatible with any remote ( S3 compatible or file based). 
Full backups creates one file per backup with all the data , that can go over 64GB, even when using xcp-ng zstd encryption
```

thomas-dkmt committed:
backups.md: Explain further how AES-256-GCM works

Add requirements on block-based remotes
Warn that encryption can't be saved if remote contains a backup
Explain how authentication works with AES-256-GCM

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
Update docs/docs/backups.md

Co-authored-by: Florent BEAUCHAMP <florent.beauchamp@vates.tech>

thomas-dkmt committed:
backups.md: Minor spelling corrections

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

thomas-dkmt committed:
backups.md: Move auth info so it applies to both algorithms

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,46 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+Currently, backups use the [`AES-256-GCM`](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) encryption algorithm. While this is a highly secure option, it does have a file size limitation of 64 GiB. This isn't an issue when working with incremental backups, as the data is split into smaller blocks, making it fully compatible with any remote (S3-compatible or file-based). 
+Full backups create one file per backup with all the data, that can go over 64 GB, even when using XCP-ng zstd encryption.
+
+
```
```suggestion

Full backups create one file per backup with all the data, that can go over 64 GB, even when using XCP-ng zstd encryption.

```

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,46 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+3. In the subsection called **Encrypt all new data sent to this remote** you will find a text area. Enter your encryption key there.
+4. Click the **Save configuration** Button to finish the encryption setup.
+
+### Current Encryption Algorithm: AES-256-GCM
```
- formatting
- change current/future to previous/current because the change will come rather soon, so that we don't have to update the doc in a week
- only keep the algo name in the title to make anchor links stable (i.e. not change in the future)
- explicit the date

```suggestion
### `AES-256-GCM`

> This algorithm was the default before February 2025 and has now been replaced by [`ChaCha20-Poly1305`](#chacha20-poly1305).
```

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,46 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+The `AES-256-GCM` algorithm is fully compliant with [ANSSI guidelines (in French)](https://cyber.gouv.fr/sites/default/files/2021/03/anssi-guide-selection_crypto-1.0.pdf).
+
+### Upcoming Change: `ChaCha20-Poly1305`
```
Make anchor link more stable.

```suggestion
### `ChaCha20-Poly1305`
```

julien-f commented on docs/docs/backups.md:
```diff
@@ -26,6 +26,46 @@ Each backups' job execution is identified by a `runId`. You can find this `runId
+
+To improve flexibility and performance, Xen Orchestra will transition to the [`ChaCha20-Poly1305`](https://en.wikipedia.org/wiki/ChaCha20-Poly1305) encryption algorithm by February 2025. This update addresses the file size limitations of `AES-256-GCM` while maintaining a high level of security and compliance with ANSSI guidelines.
+
+Backup repositories that were encrypted with `AES-256-GCM` will remain accessible, to ensure a smooth transition.
```
As it's now the new default, I believe this section should be above the legacy one.

julien-f reviewed (commented):
A couple of small comments from me, otherwise very good :)

thomas-dkmt committed:
Update docs/docs/backups.md

Co-authored-by: Julien Fontanet <julien.fontanet@vates.tech>

thomas-dkmt committed:
Update docs/docs/backups.md

Co-authored-by: Julien Fontanet <julien.fontanet@vates.tech>

thomas-dkmt committed:
Update docs/docs/backups.md

Co-authored-by: Julien Fontanet <julien.fontanet@vates.tech>

thomas-dkmt committed:
backups.md: Move ChaCha20 algo before AES-256 algo

Moving the ChaCha20-Poly1305 in first position, before
AES-256-GCM, as it's the new standard encryption algorithm

Signed-off-by: Thomas Moraine <thomas.moraine@vates.tech>
